tests/core/test_graph.py

```diff
@@ -1,54 +1,367 @@
+    g_u = simple_graph.to_undirected()
+    assert g_u.data.is_undirected()
 
-def test_N(simple_graph):
-    assert simple_graph.N == 3
 
 def test_weighted_graph(simple_graph_multi_edges):
     assert simple_graph_multi_edges.M == 4
     weighted_graph = simple_graph_multi_edges.to_weighted_graph()
-    assert weighted_graph.M == 3
-    assert weighted_graph.N == 3
-    assert weighted_graph['edge_weight', 'a', 'b'] == 2
+    assert weighted_graph.data.num_edges == 3
+    assert weighted_graph.data.num_nodes == 3
+    assert weighted_graph["edge_weight", "a", "b"] == 2
 
-def test_undirected_graph(simple_graph):
-    g_u = simple_graph.to_undirected()
-    assert g_u.is_edge('a', 'b')
-    assert g_u.is_edge('b', 'a')
-    assert g_u.is_edge('b', 'c')
-    assert g_u.is_edge('c', 'b')
-    assert g_u.is_edge('a', 'c')
-    assert g_u.is_edge('c', 'a')
 
-def test_M(simple_graph):
-    assert simple_graph.M == 3
+def test_attr_types(simple_graph):
+    attr_types = Graph.attr_types(simple_graph.data.to_dict())
+    assert isinstance(attr_types, dict)
 
 
-def test_node_attr(simple_graph):
+def test_node_attrs(simple_graph):
+    node_attrs = simple_graph.node_attrs()
+    assert isinstance(node_attrs, list)
+    assert len(node_attrs) == 0
+
     simple_graph.data["node_class"] = torch.tensor([[1], [2], [3]])
-    assert simple_graph["node_class", "a"].item() == 1
+    node_attrs = simple_graph.node_attrs()
+    assert len(node_attrs) == 1
 
 
-def test_edge_attr(simple_graph):
+def test_edge_attrs(simple_graph):
+    edge_attrs = simple_graph.edge_attrs()
+    assert isinstance(edge_attrs, list)
+    assert len(edge_attrs) == 0
+
     simple_graph.data["edge_weight"] = torch.tensor([[1], [1], [2]])
-    assert simple_graph["edge_weight", "a", "b"].item() == 1
+    edge_attrs = simple_graph.edge_attrs()
+    assert len(edge_attrs) == 1
 
 
-def test_graph_attr(simple_graph):
-    simple_graph.data["graph_feature"] = torch.tensor([42])
-    assert simple_graph["graph_feature"].item() == 42
+def test_nodes(simple_graph):
+    i = 0
+    for node in simple_graph.nodes:
+        assert node in ["a", "b", "c"]
+        i += 1
+    assert i == 3
+
+
+def test_edges(simple_graph):
+    i = 0
+    for edge in simple_graph.edges:
+        assert edge in [("a", "b"), ("b", "c"), ("a", "c")]
+        i += 1
+    assert i == 3
 
 
 def test_successors(simple_graph):
     s = [v for v in simple_graph.successors("a")]
     assert s == ["b", "c"]
 
+    s = [v for v in simple_graph.successors("c")]
+    assert len(s) == 0
+
 
 def test_predecessors(simple_graph):
     s = [v for v in simple_graph.predecessors("b")]
     assert s == ["a"]
 
+    s = [v for v in simple_graph.predecessors("a")]
+    assert len(s) == 0
+
 
 def test_is_edge(simple_graph):
     assert simple_graph.is_edge("a", "b")
+    assert not simple_graph.is_edge("b", "a")
+    assert simple_graph.is_edge("a", "c")
+    assert not simple_graph.is_edge("c", "a")
```

Comment:
TODO: Needs fixing

Reply:
This issue is due to the CSR format returned by pyG.EdgeIndex for matrix with rows that only contain zeros. 

My understanding of the CSR format is that a zero row would be represented by a repeated entry in the `row_ptr`, which allows to recover the zero row. The `get_csr` function of `EdgeIndex` seems to omits this. This also seems to be an issue when calling the `EdgeIndex.to_dense()` function, which returns a dense, possibly non-square matrix where all zero rows are missing, i.e. for the example above it returns 

```
tensor([[0., 1., 1.],
        [0., 0., 1.]])
```

instead of 

```
tensor([[0., 1., 1.],
        [0., 0., 1.],
        [0., 0., 0]])
```

Need to check whether this behavior of the pyG CSR format is intended.